python/ray/util/client/dataclient.py

```diff
@@ -35,6 +34,11 @@ def __init__(self, channel: "grpc._channel.Channel", client_id: str,
         self.ready_data: Dict[int, Any] = {}
         self.cv = threading.Condition()
         self.lock = threading.RLock()
+
+        # NOTE: Dictionary insertion is guaranteed to complete before lookup
+        # and/or removal because of synchronization via the request_queue.
+        self.asyncio_waiting_data: Dict[int, Callable[
+            [ray_client_pb2.DataResponse], None]] = {}
```

Comment:
@simon-mo I couldn't find official Pydocs on Dictionaries + Thread Safety, but IIRC python is okay with multi-threaded access as long as operations on the same key do not have race conditions.

Reply:
oh.. it's definitely protected by GIL. dict.pop() is threadsafe i believe